### PR TITLE
Billing unit tests

### PR DESCRIPTION
**Pull request by milo-hyben**
This PR contains:
1. BqTest class introduced to allow for BigQuery mocking.
2. Unit tests for all the billing python code, bringing the total metamist code coverage to 76%.
3. Members functions in billing were converted to static method where possible.
4. A few variables were renamed

### CODE REVIEW

File: test/test_bq_billing_base.py [hunk 1]
```diff
@@ -0,0 +1,885 @@
...
+        query = BillingTotalCostQueryModel(
+            fields=[],
+            start_date='2024-01-01',
+            end_date='2024-01-01',
+            filters={BillingColumn.LABELS: {'test_key': 'test_value'}},
+        )
+
+        func_filter = BillingBaseTable._prepare_labels_function(query)
+
+        self.assertEqual('getLabelValue', func_filter.func_name)
```
Comment (illusional) [inline test/test_bq_billing_base.py:350]: Are these tests still relevant with the label changes recently?

**Review by illusional — COMMENTED** [review]
Broadly looking good, are there changes to the label bits that need to be made with the table change?

I dropped a few comments in the middle, but can you split out tests that are testing multiple things (unless it's really small like testing None + 1 value), but if it requires a new set of mocking, then IMO a new test it best.

I think it's worth catching and validating queries if the instance is available, mostly just adding some auth into the GitHub action.

File: db/python/tables/bq/billing_filter.py [hunk 2]
```diff
@@ -46,3 +47,17 @@ class BillingFilter(GenericBQFilterModel):
...
+
+    def __eq__(self, other: Any) -> bool:
+        """Equality operator"""
+        result = super().__eq__(other)
+        if not result or not isinstance(other, BillingFilter):
+            return False
+
+        # compare all attributes
+        for att in self.__dict__:
+            if not getattr(self, att).__eq__(getattr(other, att)):
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply db/python/tables/bq/billing_filter.py]: Make sense.

File: test/test_api_billing.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+    BillingCostBudgetRecord,
+    BillingHailBatchCostRecord,
+    BillingTotalCostQueryModel,
+    BillingTotalCostRecord,
+)
+
+TEST_API_BILLING_USER = 'test_user'
+
+
+class TestApiBilling(BqTest):
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_api_billing.py:21]: Good idea, make one default on the top.

File: test/testbqbase.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from db.python.gcp_connect import BqConnection
+from db.python.layers.billing import BillingLayer
+
+
+class BqTest(unittest.TestCase):
+    """Base class for Big Query integration tests"""
+
+    # store connections here, so they can be created PER-CLASS
+    # and don't get recreated per test.
+
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/testbqbase.py]: Removed

File: test/test_api_utils.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    def test_get_invoice_month_range(self):
+        """
+        Test get_invoice_month_range function
+        """
+        jan_2021 = datetime.strptime('2021-01-10', '%Y-%m-%d').date()
+        res_jan_2021 = get_invoice_month_range(jan_2021)
+
+        # there is 3 (INVOICE_DAY_DIFF) days difference between invoice month st and end
+        self.assertEqual(
+            (datetime(2020, 12, 29).date(), datetime(2021, 2, 3).date()),
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_api_utils.py]: Yes, of course :-)

File: test/test_bq_billing_base.py [hunk 6]
```diff
@@ -0,0 +1,885 @@
...
+)
+
+
+def mock_execute_query(query, *_args, **_kwargs):
+    """
+    This is a mockup function for _execute_query function
+    This returns one mockup BQ query result for various SQL queries
+    This is used in a few tests
+    """
+    if 'month.cost as monthly_cost' in query:
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_bq_billing_base.py:36]: Refactored to two separate functions. 
One of the mockup function stills needs if statement as BQ is called twice in one Api call and you can only have one mockup function.

File: test/test_bq_billing_base.py [hunk 7]
```diff
@@ -0,0 +1,885 @@
...
+        # test with one empty
+        mock_rows = mock.MagicMock(spec=bq.table.RowIterator)
+        mock_rows.total_rows = 1
+        mock_rows.__iter__.return_value = [{}]
+
+        query_job_result = mock.MagicMock(spec=bq.job.QueryJob)
+        query_job_result.result.return_value = mock_rows
+
+        single_row = BillingBaseTable._convert_output(query_job_result)
+        self.assertEqual([{}], single_row)
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_bq_billing_base.py]: Have split them as suggested.

File: test/test_bq_billing_gcp_daily.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+            '        )\n'
+            '        \n'
+            '            GROUP BY\n'
+            '                field,\n'
+            '                cost_category\n'
+            '        ) day\n'
+            '        ON month.field = day.field\n'
+            '        AND month.cost_category = day.cost_category\n'
+            '        '
+        )
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_bq_billing_gcp_daily.py]: Very good point!

File: test/test_bq_billing_base.py [hunk 9]
```diff
@@ -0,0 +1,885 @@
...
+        ]
+        mock_result = mock.MagicMock(spec=bq.job.QueryJob)
+        mock_result.result.return_value = mock_rows
+        return mock_result
+
+    # otherwise no results
+    return []
+
+
+class TestBillingBaseTable(BqTest):
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_bq_billing_base.py:74]: Regarding Cost of query I will create a new PR for that.
Having minimal BQ metadata similar is a good idea, however it start smelling as integration tests :-)
Any I will put it on the list.

File: test/test_bq_billing_base.py (hunk 1 above)
Comment (milo-hyben) [reply test/test_bq_billing_base.py:374]: Valid point, there are not used after we made change to the BQ table layout, 
let me remove all those labelFunction tests.

**Comment by milo-hyben** [comment]
Hi Michael, I had addresses most of your comment. The only left for the future PR is to use local BQ installation to run those tests against BQ tables. As well cost of query I will add to the response in a different PR.

**Review by illusional — APPROVED** [review]
😮‍💨

File: test/test_api_billing.py [hunk 10]
```diff
@@ -14,6 +14,9 @@
 
 TEST_API_BILLING_USER = 'test_user'
 
+# make billing enabled by default
+patch('api.routes.billing.is_billing_enabled', return_value=True)
+
 
 class TestApiBilling(BqTest):
```
[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_api_billing.py:21]: I guess it is treated as a function call and not a decorator.

[the remark this answers is not in the document]
Comment (illusional) [reply test/test_api_billing.py]: Oh it's supposed to have a `@` in front of it!!

[the remark this answers is not in the document]
Comment (milo-hyben) [reply test/test_api_billing.py]: Nope, that can only be done in front of a function call. I have tried it did not work for the class.

**Pull request merged**